### PR TITLE
feat: database module

### PR DESCRIPTION
**Pull request by Salakar**
This PR implements support for database endpoints and additionally serving functionality for the admin api + an example functions project.

## Testing

I have a bunch of unit tests stashed but they require a bit of work since latest refactors/changes so I've not pushed these. In light of time constraints I've focused on manual end-to-end testing instead, results below;

**Deployed example**:
```python
"""
Example Firebase Functions written in Python
"""
from firebase_functions import db, options

options.set_global_options(
    region=options.SupportedRegion.EUROPE_WEST1,
    memory=options.MemoryOption.MB_256,
)


@db.on_value_written(
    reference="hello",
    region=options.SupportedRegion.EUROPE_WEST1,
)
def on_write_example(event: db.DatabaseEvent[db.Change[object]]) -> None:
    print("Hello from db write event:", event)


@db.on_value_created(reference="hello/{any_thing_here}/bar")
def on_created_example(event: db.DatabaseEvent[object]) -> None:
    print("Hello from db create event:", event)


@db.on_value_deleted(reference="hello/{any_thing_here}/bar")
def on_deleted_example(event: db.DatabaseEvent[object]) -> None:
    print("Hello from db delete event:", event)


@db.on_value_updated(reference="hello")
def on_updated_example(event: db.DatabaseEvent[db.Change[object]]) -> None:
    print("Hello from db updated event:", event)
```

**Generated manifest yaml**:
```yaml
endpoints:
  on-created-example:
    availableMemoryMb: 256
    cpu: gcf_gen1
    entryPoint: on_created_example
    eventTrigger:
      eventFilterPathPatterns:
        instance: '*'
        ref: hello/{any_thing_here}/bar
      eventFilters: {}
      eventType: google.firebase.database.ref.v1.created
      retry: false
    platform: gcfv2
    region:
    - europe-west1
    secretEnvironmentVariables: []
  on-deleted-example:
    availableMemoryMb: 256
    cpu: gcf_gen1
    entryPoint: on_deleted_example
    eventTrigger:
      eventFilterPathPatterns:
        instance: '*'
        ref: hello/{any_thing_here}/bar
      eventFilters: {}
      eventType: google.firebase.database.ref.v1.deleted
      retry: false
    platform: gcfv2
    region:
    - europe-west1
    secretEnvironmentVariables: []
  on-updated-example:
    availableMemoryMb: 256
    cpu: gcf_gen1
    entryPoint: on_updated_example
    eventTrigger:
      eventFilterPathPatterns:
        instance: '*'
        ref: hello
      eventFilters: {}
      eventType: google.firebase.database.ref.v1.updated
      retry: false
    platform: gcfv2
    region:
    - europe-west1
    secretEnvironmentVariables: []
  on-write-example:
    availableMemoryMb: 256
    cpu: gcf_gen1
    entryPoint: on_write_example
    eventTrigger:
      eventFilterPathPatterns:
        instance: '*'
        ref: hello
      eventFilters: {}
      eventType: google.firebase.database.ref.v1.written
      retry: false
    platform: gcfv2
    region:
    - europe-west1
    secretEnvironmentVariables: []
params: []
specVersion: v1alpha1
```

**All database function types deploy successfully**:
image

**Logs from live deployment showing successful invocations**:
```
2022-10-09 01:54:55.673 BSTHello from db write event: DatabaseEvent(specversion='1.0', id='eG8EgvQPooVBLfc1PiO2E5waVFo=', source='//firebasedatabase.googleapis.com/projects/_/locations/europe-west1/instances/python-functions-testing-default-rtdb', type='google.firebase.database.ref.v1.written', time=datetime.datetime(2022, 10, 9, 0, 54, 53, 582000, tzinfo=datetime.timezone.utc), data=Change(before={'foo': 'barrrr', 'testttt': 123, 'world': True}, after={'foo': 'bar', 'testttt': 123, 'world': True}), subject='refs/hello', firebase_database_host='europe-west1.firebasedatabase.app', instance='python-functions-testing-default-rtdb', reference='hello', location='europe-west1', params={})
2022-10-09 03:21:10.879 BSTHello from db create event: DatabaseEvent(specversion='1.0', id='re5ZZnJCTRxS82I2vbwcZdc6hI4=', source='//firebasedatabase.googleapis.com/projects/_/locations/europe-west1/instances/python-functions-testing-default-rtdb', type='google.firebase.database.ref.v1.created', time=datetime.datetime(2022, 10, 9, 2, 21, 7, 77000, tzinfo=datetime.timezone.utc), data=Change(before=None, after=111), subject='refs/hello/anything/bar', firebase_database_host='europe-west1.firebasedatabase.app', instance='python-functions-testing-default-rtdb', reference='hello/anything/bar', location='europe-west1', params={})
2022-10-09 03:24:04.618 BSTHello from db delete event: DatabaseEvent(specversion='1.0', id='3MkafdR7RV0REVyJ887AU1yhdig=', source='//firebasedatabase.googleapis.com/projects/_/locations/europe-west1/instances/python-functions-testing-default-rtdb', type='google.firebase.database.ref.v1.deleted', time=datetime.datetime(2022, 10, 9, 2, 24, 2, 60000, tzinfo=datetime.timezone.utc), data=Change(before=111, after=None), subject='refs/hello/anything/bar', firebase_database_host='europe-west1.firebasedatabase.app', instance='python-functions-testing-default-rtdb', reference='hello/anything/bar', location='europe-west1', params={})
2022-10-09 01:54:56.333 BSTHello from db updated event: DatabaseEvent(specversion='1.0', id='dNfxxrJEahk8Q8VDSNLmXCQehTI=', source='//firebasedatabase.googleapis.com/projects/_/locations/europe-west1/instances/python-functions-testing-default-rtdb', type='google.firebase.database.ref.v1.updated', time=datetime.datetime(2022, 10, 9, 0, 54, 53, 582000, tzinfo=datetime.timezone.utc), data=Change(before={'foo': 'barrrr', 'testttt': 123, 'world': True}, after={'foo': 'bar', 'testttt': 123, 'world': True}), subject='refs/hello', firebase_database_host='europe-west1.firebasedatabase.app', instance='python-functions-testing-default-rtdb', reference='hello', location='europe-west1', params={})
```

### CODE REVIEW

File: src/firebase_functions/db.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+_C2 = _typing.Callable[[_E2], None]
+
+
+def _db_endpoint_handler(
+    func: _C1 | _C2,
+    raw: _ce.CloudEvent,
+) -> None:
+    event_attributes = raw._get_attributes()
+    event_data = raw._get_data()
+    # TODO Params are built locally via path pattern which is currently unimplemented
```
Comment (Salakar) [inline src/firebase_functions/db.py:66]:
```suggestion
    # TODO Params are built locally via path pattern which is currently unimplemented
```

Extracting path patterns into params isn't currently implemented, you can still define patterns in your ref or instance names and they work, just no automatic `params` dict generation yet. I will look into this after `https` support due to time constraints.

File: src/firebase_functions/private/serving.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        endpoints[name.replace("_", "-").lower()] = endpoint
+
+    manifest_stack = _manifest.ManifestStack(endpoints=endpoints)
+    manifest_spec = _manifest.manifest_to_spec_dict(manifest_stack)
+    manifest_spec_with_sentinels = to_spec(manifest_spec)
+
+    def represent_sentinel(self, _):
+        # TODO distinguishing between RESET_VALUE or DEFAULT_VALUE
+        # TODO can be done here
+        return self.represent_scalar("tag:yaml.org,2002:null", "null")
```
Comment (Salakar) [inline src/firebase_functions/private/serving.py:68]: Note: Should be able to add RESET_VALUE support here, will do in a separate PR after `https`

File: src/firebase_functions/db.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    )
+    func(database_event)
+
+
+@_util.copy_func_kwargs(_options.DatabaseOptions)
+def on_value_written(**kwargs) -> _typing.Callable[[_C1], _C1]:
+    """
+    Event handler which triggers when data is created, updated, or deleted in Realtime Database.
+    Example::
+      @on_value_written(reference="*")
```
Comment (inlined) [inline src/firebase_functions/db.py:104]: nit: "*" makes a bad reference example. If someone runs it and there's a large database they could impact production and it doesn't show off parameters which are kinda the whole point. Maybe something like `"messages/{id}"` or `"users/{uid}"`

File: src/firebase_functions/db.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+    params: dict[str, str]
+    """
+    An dict containing the values of the path patterns.
+    Only named capture groups are populated - {key}, {key=*}, {key=**}
+    """
+
+
+_E1 = DatabaseEvent[Change[_typing.Any | None]]
+_E2 = DatabaseEvent[_typing.Any | None]
```
Comment (inlined) [inline src/firebase_functions/db.py:55]: How can the payload ever be `None`?

File: src/firebase_functions/db.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            "%Y-%m-%dT%H:%M:%S.%f%z",
+        ),
+        data=database_event_data,
+        subject=event_attributes["subject"],
+        params=params,
+    )
+    func(database_event)
+
+
+@_util.copy_func_kwargs(_options.DatabaseOptions)
```
Comment (inlined) [inline src/firebase_functions/db.py:99]: Also, _if_ we can make DatabaseOptions work, we should put all Database stuff in the db package.

File: src/firebase_functions/db.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
+"""
+Module for Cloud Functions that are triggered by the Firebase Realtime Database.
+"""
+# pylint: disable=protected-access
+import dataclasses as _dataclass
```
Comment (inlined) [inline src/firebase_functions/db.py:5]: Please follow the Google Python style guide, including its [guidance for imports](https://google.github.io/styleguide/pyguide.html#22-imports)

File: src/firebase_functions/options.py [hunk 7]
```diff
@@ -223,32 +197,141 @@ def asdict_with_global_options(self) -> dict:
                 merged_options[key] = global_options[key]
         # None values are automatically stripped out in ManifestEndpoint generation.
```
Comment (inlined) [inline src/firebase_functions/options.py:198]: About that... we should sync. In the Node SDK we're now defaulting to using nulls for all unspecified options if a new option, preserveExistingChanges, is set to false (default). We will draft an API proposal for you and send it out, but you can assume it will pass. No need to make that change in this PR, just want to give a heads up.

File: src/firebase_functions/options.py [hunk 8]
```diff
@@ -223,32 +197,141 @@ def asdict_with_global_options(self) -> dict:
...
+                        secret_value = secret.name
+                    return {"key": secret_value}
+
+                secret_envs = list(
+                    map(convert_secret, _typing.cast(list, options.secrets)))
+            elif options.secrets is _util.Sentinel:
+                secret_envs = _typing.cast(_util.Sentinel, options.secrets)
+
+        region: list[str] | None = None
+        if isinstance(options.region, list):
```
Comment (inlined) [inline src/firebase_functions/options.py:228]: subtlety note: we allow region to be a list for HTTPS functions but no other functions. Yet in the functions.yaml it will always be a list for simplicity.

File: src/firebase_functions/options.py [hunk 9]
```diff
@@ -223,32 +197,141 @@ def asdict_with_global_options(self) -> dict:
...
+
     topic: str
     """
     The Pub/Sub topic to watch for message events.
     """
 
-    retry: _typing.Optional[bool] = None
+
+@_dataclasses.dataclass(frozen=True, kw_only=True)
+class DatabaseOptions(RuntimeOptions):
```
Comment (inlined) [inline src/firebase_functions/options.py:280]: why `options.DatabaseOptions` instead of `db.Options`? (Again, not sure we can use such a class since it may screw up pydocs)

File: src/firebase_functions/private/serving.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
+    else:
+        raise Exception(
+            "Firebase Functions for Python could not find the main.py file in your project."
+        )
+    functions = inspect.getmembers(module, inspect.isfunction)
+    firebase_functions = {}
+    for entry in functions:
+        if hasattr(entry[1], "__firebase_endpoint__"):
```
Comment (inlined) [inline src/firebase_functions/private/serving.py:32]: Forgive my ignorance. Why is entry being subscripted? Is this worth leaving a comment to explain?

File: src/firebase_functions/private/serving.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        (k, convert_value(v)) for k, v in data.items() if v is not None)
+    return without_nones
+
+
+def functions_as_yaml(functions: dict) -> str:
+    endpoints: dict[str, _manifest.ManifestEndpoint] = {}
+    for name, function in functions.items():
+        endpoint = function.__firebase_endpoint__
+        # v2 function name(s) can only contain lower case letters, numbers, hyphens
+        endpoints[name.replace("_", "-").lower()] = endpoint
```
Comment (inlined) [inline src/firebase_functions/private/serving.py:59]: Let's not do this. We considered changing names to be v2 compatible but
1. We'd have to do some rearchitecting of the Firebase CLI to not interpret "-" as a grouping separator
2. GCF is looking at adding back support for these special characters and that will ultimately bet he better solution

For now, you can just not do the transform and our CLI will throw that the name is invalid. Just like JS customers have to write asinglewordwithnocapitalization python customers will temporarily have to do the same.

File: src/firebase_functions/private/serving.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    manifest_stack = _manifest.ManifestStack(endpoints=endpoints)
+    manifest_spec = _manifest.manifest_to_spec_dict(manifest_stack)
+    manifest_spec_with_sentinels = to_spec(manifest_spec)
+
+    def represent_sentinel(self, _):
+        # TODO distinguishing between RESET_VALUE or DEFAULT_VALUE
+        # TODO can be done here
```
Comment (inlined) [inline src/firebase_functions/private/serving.py:67]: What difference between RESET_VALUE and DEFAULT_VALUE? Sorry, I remember that we had two uses of sentinel but I can't recall what the second case was and those two names sound unfortunately close together.

File: src/firebase_functions/private/serving.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        "/__/quitquitquit",
+        endpoint="quitquitquit",
+        view_func=quitquitquit,
+    )
+
+    return app
+
+
+def main():
+    if os.environ["ADMIN_PORT"] is not None:
```
Comment (inlined) [inline src/firebase_functions/private/serving.py:109]: The port env var is just `PORT` (so that it works with Run)

File: src/firebase_functions/private/serving.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        kill(getpid(), SIGTERM)
+
+    response = Response("OK", status=200)
+    response.call_on_close(quit_after_close)
+    return response
+
+
+def serve_admin() -> Flask:
+    app = Flask(__name__)
+    app.add_url_rule(
```
Comment (inlined) [inline src/firebase_functions/private/serving.py:93]: Only add these rules if FUNCTIONS_CONTROL_API is "true" (or truthy if you care)

**Review by inlined — CHANGES_REQUESTED** [review]

File: src/firebase_functions/core.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    """
+
+    subject: str | None
+    """
+    The resource, provided by source, that this event relates to
+    """
+
+
+@_dataclass.dataclass(frozen=True)
+class Change(_typing.Generic[T]):
```
[the remark this answers is not in the document]
Comment (Salakar) [reply src/firebase_functions/core.py:55]: Context; I mirrored to what v2 functions js does; https://github.com/firebase/firebase-functions/blob/master/src/v2/core.ts#L70 - since `Change` can be used by Firestore also (from my understanding)

File: src/firebase_functions/db.py (hunk 6 above)
Comment (Salakar) [reply src/firebase_functions/db.py:5]: The only reason I've not done this is that otherwise these imports pollute the exports of the module when users import it, e.g. 

Without `_`:

image



With `_`: 

image


This unofficial convention is followed in a few google projects including the python admin sdk; https://github.com/firebase/firebase-admin-python/blob/master/firebase_admin/db.py

If we're ok with having all imports appear in exports of the modules I can change this back?

An alternative to prevent exports pollution is tp define an `__all__` in each file which is a list of string names of things that should be exported - but this is often overlooked when updates happen since it's a manually created list of string names of module members.

File: src/firebase_functions/db.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Salakar) [reply src/firebase_functions/db.py:99]: I can confirm code completion/autosense in vscode works like this;

image
image

It's worth noting that I've probably not named this fn right, it doesn't copy anything, it's more `infer_kwargs(source_func)` its purely a types util so anything relying on types just works (e.g. vscode completions) - I'll check pydocs tomorrow to confirm though
